packages/firestore/test/integration/api/bundle.test.ts

```diff
@@ -0,0 +1,257 @@
+            progresses.push(progress);
+          },
+          error => {},
+          progress => {
+            completeProgress = progress!;
+          }
+        );
+
+      // No loading actually happened in the second `loadBundle` call.
+      expect(progresses).to.be.empty;
+      verifySuccessProgress(completeProgress!);
+
+      // Read from cache. These documents do not exist in backend, so they can
+      // only be read from cache.
+      const snap = await db.collection('coll-1').get({ source: 'cache' });
+      expect(toDataArray(snap)).to.deep.equal([
+        { k: 'a', bar: 1 },
+        { k: 'b', bar: 2 }
+      ]);
+    });
+  });
+
+  it('load with documents already pulled from backend.', () => {
+    return withTestDb(persistence, async db => {
+      await db.doc('coll-1/a').set({ k: 'a', bar: 0 });
+      await db.doc('coll-1/b').set({ k: 'b', bar: 0 });
+
+      const accumulator = new EventsAccumulator<firestore.QuerySnapshot>();
+      db.collection('coll-1').onSnapshot(accumulator.storeEvent);
+      await accumulator.awaitEvent();
+
+      const builder = bundleWithTestDocs(db);
+      const progress = await db.loadBundle(
+        // Testing passing non-string bundles.
+        encoder.encode(
+          builder.build('test-bundle', { seconds: 1001, nanos: 9999 })
+        )
+      );
+
+      verifySuccessProgress(progress);
+      // The test bundle is holding ancient documents, so no events are
+      // generated as a result. The case where a bundle has newer doc than
+      // cache can only be tested in spec tests.
+      await accumulator.assertNoAdditionalEvents();
+
+      const snap = await db.collection('coll-1').get();
+      expect(toDataArray(snap)).to.deep.equal([
+        { k: 'a', bar: 0 },
+        { k: 'b', bar: 0 }
+      ]);
+    });
+  });
+
+  it('load with documents from other projects fails.', () => {
+    return withTestDb(persistence, async db => {
+      let builder = bundleWithTestDocs(db);
+      return withAlternateTestDb(persistence, async otherDb => {
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        expect(
+          otherDb.loadBundle(
+            builder.build('test-bundle', { seconds: 1001, nanos: 9999 })
+          )
+        ).to.be.rejectedWith('Tried to deserialize key from different project');
+
+        // Verify otherDb still functions, despite loaded a problematic bundle.
+        builder = bundleWithTestDocs(otherDb);
+        const progress = await otherDb.loadBundle(
+          builder.build('test-bundle', { seconds: 1001, nanos: 9999 })
+        );
+        verifySuccessProgress(progress);
+
+        // Read from cache. These documents do not exist in backend, so they can
+        // only be read from cache.
+        const snap = await otherDb
+          .collection('coll-1')
+          .get({ source: 'cache' });
+        expect(toDataArray(snap)).to.deep.equal([
+          { k: 'a', bar: 1 },
+          { k: 'b', bar: 2 }
```

Comment:
Does this add any additional coverage for this test? Wouldn't `otherDb.loadBundle` succeeding just be enough?